scene/2d/parallax_2d.cpp

```diff
@@ -0,0 +1,277 @@
+
+			real_t mod = Math::fmod(autoscroll_offset.x, repeat_size.x);
+			autoscroll_offset.x = mod < 0 ? mod + repeat_size.x : mod;
+			mod = Math::fmod(autoscroll_offset.y, repeat_size.y);
+			autoscroll_offset.y = mod < 0 ? mod + repeat_size.y : mod;
+
+			_update_scroll();
+		} break;
+
+		case NOTIFICATION_ENTER_TREE: {
+			group_name = "__cameras_" + itos(get_viewport_rid().get_id());
+			add_to_group(group_name);
+			_update_repeat();
+		} break;
+
+		case NOTIFICATION_EXIT_TREE: {
+			remove_from_group(group_name);
+		} break;
+	}
+}
+
+void Parallax2D::_camera_moved(const Transform2D &p_transform, const Point2 &p_screen_offset, const Point2 &p_adj_screen_pos) {
+	if (ignore_camera_scroll) {
+		return;
+	}
+
+	set_screen_offset(p_adj_screen_pos);
+}
+
+void Parallax2D::_update_scroll() {
+	if (!is_inside_tree() || Engine::get_singleton()->is_editor_hint()) {
+		return;
+	}
+
+	Point2 scroll_ofs = screen_offset;
+	Size2 vps = get_viewport_rect().size;
+
+	if (limit_begin.x <= limit_end.x - vps.x) {
+		if (screen_offset.x < limit_begin.x) {
+			scroll_ofs.x = limit_begin.x;
+		} else if (screen_offset.x + vps.x > limit_end.x) {
+			scroll_ofs.x = limit_end.x - vps.x;
+		}
+	}
+
+	if (limit_begin.y <= limit_end.y - vps.y) {
+		if (screen_offset.y < limit_begin.y) {
+			scroll_ofs.y = limit_begin.y;
+		} else if (screen_offset.y + vps.y > limit_end.y) {
+			scroll_ofs.y = limit_end.y - vps.y;
+		}
+	}
+
+	scroll_ofs *= scroll_scale;
+
+	if (repeat_size.x) {
+		real_t mod = Math::fmod(scroll_ofs.x + offset.x + autoscroll_offset.x, repeat_size.x);
+		mod = mod < 0 ? mod + repeat_size.x : mod;
+		scroll_ofs.x = screen_offset.x - mod;
+	} else {
+		scroll_ofs.x = screen_offset.x + offset.x - scroll_ofs.x;
+	}
+
+	if (repeat_size.y) {
+		real_t mod = Math::fmod(scroll_ofs.y + offset.y + autoscroll_offset.y, repeat_size.y);
+		mod = mod < 0 ? mod + repeat_size.y : mod;
+		scroll_ofs.y = screen_offset.y - mod;
+	} else {
+		scroll_ofs.y = screen_offset.y + offset.y - scroll_ofs.y;
+	}
+
+	if (!follow_viewport) {
+		scroll_ofs -= screen_offset;
+	}
+
+	set_global_position(scroll_ofs);
+}
+
+void Parallax2D::_update_repeat() {
+	if (!is_inside_tree()) {
+		return;
+	}
+
+	RID ci = get_canvas_item();
+	Point2 repeatScale = repeat_size * get_scale();
+	RenderingServer::get_singleton()->canvas_set_item_repeat(ci, repeatScale, repeat_times);
+}
+
+void Parallax2D::set_scroll_scale(const Size2 &p_scale) {
+	scroll_scale = p_scale;
+}
+
+Size2 Parallax2D::get_scroll_scale() const {
+	return scroll_scale;
+}
+
+void Parallax2D::set_repeat_size(const Size2 &p_repeat_size) {
+	repeat_size = p_repeat_size;
+
+	if (repeat_size.x < 0) {
+		repeat_size.x = 0;
+	}
+
+	if (repeat_size.y < 0) {
+		repeat_size.y = 0;
+	}
+
+	_update_repeat();
+}
+
+Size2 Parallax2D::get_repeat_size() const {
+	return repeat_size;
+}
+
+void Parallax2D::set_repeat_times(int p_repeat_times) {
+	repeat_times = p_repeat_times;
+
+	if (repeat_times < 1) {
+		repeat_times = 1;
+	}
```

Comment:
Would prefer MIN thing, but it's probably not important.

Reply:
Also MAX

Reply:
I don't see a max helper method for ints?

Reply:
It's a macro, `MAX`, defined in `typedefs.h`

Reply:
Ah. Missed that. Thanks!